### DISCUSSION TITLE
useModel在TypeScript中怎么使用

### DISCUSSION THREAD

**Discussion by rangyf** [umijs/umi#7233] [Q&A]
按照 [umi官网useModel的介绍](https://umijs.org/zh-CN/plugins/plugin-model#usemodel) 和 [antd pro关于useModel的介绍](https://pro.ant.design/zh-CN/docs/simple-model) 都没看出在TypeScript情况下怎么使用useModel

比如我的models/test.ts

`
import { useState } from 'react';
export interface TestType {
  text: string
};
export default () => {
  const [data, setData] = useState<TestType | undefined>(undefined);
  return { data, setData };
};
`

然后在别处使用该model
`
// 报错提示Argument of type '"test"' is not assignable to parameter of type '"@@initialState"'.
const { data, setData } = useModel('test');
`

也不知该在泛型useModel\<T\>中写什么？

**Comment by leftstick** [comment]
参考[generator-umi](https://github.com/DFocusGroup/generator-umi)，里面有完整的利用 `useModel` 的增删改查的例子

**Reply by rangyf** [reply]
抱歉之前没看到回复…其实我贴的俩链接里面的例子也没问题…原来重点在models/test.ts文件export default不能直接用箭头函数，要用
`function test() {...}`
或者先声明一个
`const test = () => {...} `
再
`export default test`
这个export的名字才是它机制里面的重点…我之前钻牛角尖一直以为它会默认根据文件名来判断

**Reply by leftstick** [reply]
没明白你的问题。这几种都是合法的model声明方式，umi, ts都能准确识别：

**直接默认导出函数**

```ts
import { useState } from 'react'

export default function useTestModel() {
  const [txt, setTxt] = useState<string>('')

  return {
    txt,
    setTxt,
  }
}
```

**直接默认导出箭头函数**

```ts
import { useState } from 'react'

export default () => {
  const [txt, setTxt] = useState<string>('')

  return {
    txt,
    setTxt,
  }
}
```

**默认导出变量，变量为函数**

```ts
import { useState } from 'react'

function useTestModel() {
  const [txt, setTxt] = useState<string>('')

  return {
    txt,
    setTxt,
  }
}

export default useTestModel
```

**默认导出为变量，变量为箭头函数**

```ts
import { useState } from 'react'

const useTestModel = () => {
  const [txt, setTxt] = useState<string>('')

  return {
    txt,
    setTxt,
  }
}

export default useTestModel
```


上面几种写法，都能被正常解析，然后使用效果如下：

![demo](https://user-images.githubusercontent.com/1059956/132808800-ee006035-663e-4faf-bb51-87afa4f0c265.gif)

**Reply by rangyf** [reply]
这样的么，在我这里您写的第二种解析不到诶

**Reply by leftstick** [reply]
你说 『直接默认导出箭头函数』？不应该啊，你这是什么版本的 `@umijs/plugin-model` ?

**Comment by rangyf** [comment]
@***@***.***（在yarn.lock里看到的），这个项目是想学习新的ant design pro
v5的时候通过yarn create umi来的，package.json里面倒没有手动加这个dependency

Howard.Zuo ***@***.***> 于2021年9月10日周五 下午3:57写道：
